### PR TITLE
WP-804 breakdown refactor + Bug #13: Mark recorded in one zone

### PR DESCRIPTION
**Pull request by sergeychernyshev**
Fixes https://meetup.atlassian.net/browse/WP-804 / https://github.com/sergeychernyshev/ux-capture/issues/13

Also includes massive refactoring to break clever single-file prototype contraption into more manageable components.

All suggestions to avoid the awkward singleton situation (which matches window's single user timing pipeline) as well as general suggestions about refactoring in relation to bundling it into a single file to be inlined are really appreciated.

### CODE REVIEW

**Comment by sergeychernyshev** [comment]
I have a feeling I am going quite wrong on all this static function calling for singleton purposes.
Remember that @mmcgahan suggested dependency injection for subcomponents, will try to rewrite them using this approach.

**Review by mmcgahan — DISMISSED** [review]
looks like you used `npm` instead of `yarn`, so there is a `package-lock.json` file that shouldn't be there. We use `yarn` exclusively for all Meetup repos, so make sure you have an up to date install of that

**Comment by sergeychernyshev** [comment]
@mmcgahan updated to yarn

File: js/src/UXCapture.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    // assign singleton instance to window object
+    window.UX = new UXCapture(window);
+
+    _instance = window.UX;
+
+    return window.UX;
+  }
+
+  static isUserTimingSupported() {
+    return (
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: this method would `throw` if called before `_instance` was assigned a value

File: js/src/UXCapture.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+  static isConsoleTimeStampSupported() {
+    return (
+      typeof _instance.window.console !== "undefined" &&
+      typeof _instance.window.console.timeStamp !== "undefined"
+    );
+  }
+
+  constructor(window) {
+    this.window = window;
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: this creates an awkward circular reference, with `this.window.UX === this.window.UX.window.UX...`, which might explode people's brains. Is it necessary? Can we just test that we have access to a global `window`-like environment and avoid assigning `this.window` as an instance property? Essentially, I think we should crash if we're running outside of the browser (or a simulated browser environment like jsdom, which is what Jest uses).

File: js/src/UXCapture.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
+import ExpectedMark from "./ExpectedMark";
+import Zone from "./Zone";
+
+let _instance = null;
+
+export default class UXCapture {
+  static attachTo(window) {
+    // allow running in Node.js environment
+    if (typeof window === "undefined") {
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: I would argue that if a consumer passes in `undefined` to this static method, it's a user error and we should bail/throw.

File: js/src/UXCapture.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    this.onMark = null;
+    this.onMeasure = null;
+  }
+
+  static recordMeasure(zone, lastMark) {
+    if (UXCapture.isUserTimingSupported()) {
+      // record a measure using W3C User Timing API
+      _instance.window.performance.measure(
+        zone.getLabel(),
+        "navigationStart",
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: probably worth leaving a code comment that this start mark is only useful for full page load - we'll want to make this an instance property or something when we update the library to support dynamic navigation, so might as well call it out

File: js/src/Zone.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import ExpectedMark from "./ExpectedMark";
+import UXCapture from "./UXCapture";
+
+// array holding current, global set of expected zones
+let expectedZones = [];
+
+/**
+ * Zone represents collection of elements groupped together and corresponding phase of page load
```
Comment (mmcgahan) [inline js/src/Zone.js:5]: 'elements' is ambiguous here, particularly in the context of something working in the DOM, where 'element' typically means 'HTML Element' - I don't think a zone is a 'collection of html elements', though?

File: js/src/Zone.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * Only one set of zones can be tracked for the same view at one point in time
```
Comment (mmcgahan) [inline js/src/Zone.js:7]: I think it might be useful to go ahead and introduce the concept of a 'view' into code, which would be the object that groups the 'expected zones'. I think that would make the logic more clear. My understanding is that:

- one View has many Zones (typically 4)
- one Zone may have many Marks, where the 'latest' recorded mark defines the end time of the measure. Currently, all zones treat 'navigationStart' as the 'start mark'

File: js/src/UXCapture.js (hunk 4 above)
Comment (mmcgahan) [inline js/src/UXCapture.js]: I'm actually wondering if we should go ahead and require that all zones declare the name of their `start` mark, and maybe default to `'navigationStart'` for now - this line would then be `zone.start.getLabel()` or something

File: js/src/Zone.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ */
+export default class Zone {
```
Comment (mmcgahan) [inline js/src/Zone.js:9]: Zones seem to overlap quite a bit with the performance timing concept of a `measure` - do you think it might reduce confusion to call it `Measure`?

File: js/src/Zone.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  static setExpectedZones(zones) {
```
Comment (mmcgahan) [inline js/src/Zone.js]: This method feels like something that would be the responsibility of a Zone 'controller', not the `Zone` class itself - I think a `View` abstraction would be a good fit - each `View` instance would have its own set of `expectedZones`, or more simply, each `View` would have `this.zones` - you wouldn't need to keep re-assigning the `expectedZones` variable.

File: js/src/Zone.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      if (zone.marks && zone.marks.length > 0) {
+        const expectedZone = new Zone(zone.label, zone.marks);
+
+        return expectedZone;
+      }
+    });
+  }
+
+  // constructs individual zone object
+  constructor(label, marks, elements = []) {
```
Comment (mmcgahan) [inline js/src/Zone.js]: let's rename `marks` to `markLabels` just to make it clear it is not a collection of `Mark` instances.

File: js/src/ExpectedMark.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+  constructor(markLabel) {
+    this.label = markLabel;
+    this.zones = [];
+
+    this.done = null;
+
+    this.promise = new Promise((resolve, reject) => {
+      this.done = () => {
+        UXCapture.recordMark(this);
```
Comment (mmcgahan) [inline js/src/ExpectedMark.js]: I think this is one source of complexity because it's sort of circular - UXCapture tracks Zones, Zones track Marks, and Marks call UXCapture when they're 'done'. I think we could figure out a way to use callbacks passed to each class's constructor to keep each 'layer' independent. `UXCapture` could pass down its `config.onMeasure` and `config.onMark` callbacks to `Zone`, which would be responsible for calling `onMeasure` and passing `onMark` down to `Mark`, which would be responsible for calling `onMark`. That kind of setup would follow the React convention of passing down props to children and letting children use callbacks to talk to their parents rather than directly calling their parents' static methods.

**Review by mmcgahan — COMMENTED** [review]
This is a good start toward organizing the ux capture logic - I'm finding it much easier to follow. We'll talk about this when we pair later on, but I think there's also some room to streamline it a bit further.

Just to consolidate some of the ideas in individual comments:

I think we should start formally defining a `View` class, which would be responsible for managing the `Zone`s for a particular view. Each `Zone` would keep a reference to its `Mark`s. The View constructor would also be a clear place to do something like `clearMarks()`, `clearMeasures()` to re-set the marks/measures when a new view appears on dynamic navigation, but we can obviously sort that out later.

I think it might also be helpful to rename `Zone` to `Measure`, and possibly rename `label` to `name` (both for Zones and Marks) to help align the library with the terminology in the User Timing API spec - I know zones aren't quite the same as measures, but I'm not sure they're different enough to introduce new terminology.

I'm not familiar enough with the behavior of the code yet, but I'm sort of uncertain about the value of using Promises to keep track of when marks are 'recorded' vs using an `onMark` callback that would allow a Zone/Measure to know when its _last_ mark is recorded - I'd like to talk more about how that might work.

Overall I'd like to move away from having quite so many static methods that make 'controller' responsibilities a bit less clear, but I think we can do that with incremental changes.

File: js/src/Zone.js (hunk 7 above)
Comment (sergeychernyshev) [reply js/src/Zone.js:9]: Not exactly. In fact, I should probably reveal next thing about Zones - they are not really a representation of UserTiming events but rather a collection of UI elements/components/features (all names are ambiguous, unfortunately).

Each View should be loaded in phases that progressively enhance experience revealing one zone after another until full functionality is loaded.

Each phase is measured by capturing events which can be determined using some method. There are two methods I currently know of:
1. collection of Performance Timing APIs events, mostly UserTiming API marks included in the application code, but can also include events available in the browser, like "first paint", for example, or ElementTiming API. This method can be used with synthetic and RUM capture technologies as both can execute JavaScript.
2. video analysis which can determine when each element of the zone was actually painted on the screen and therefore when zone was complete. Detection of boundaries of each component can be accomplished by capturing coordinates of each HTML element zone consists of and using WPT visual progression / speed index algorithms on a modified video frames that only include pixels within those boundaries. This method can only be used with synthetic capture technologies as they require video frame recording.

Both methods have their unique advantages, here are the main ones:
1. UserTiming API can be ran in RUM and therefore is the only way to correlate results with actual business KPIs for specific users
2. Video analysis actually represents what is displayed to the user and exactly aligns with visual user experience requirements while UserTiming is an approximation of browser behavior.

Both methods are necessary to ensure successful implementation in the long run as synthetic capture can calibrate RUM results providing realistic accuracy measures.

Long story short, Zone does not really equate to measure in this library because we'll need to define both UserTiming marks and CSS selectors for zones to accomplish both.

I realized that CSS selectors (or other methods of identifying specific HTML Elements) are not really optional as we were discussing this instrumentation with @sadafie and I was thinking of how we'd work with product managers and designers to begin with.

File: js/src/Zone.js (hunk 6 above)
Comment (sergeychernyshev) [reply js/src/Zone.js:7]: I agree, it seems like a good abstraction to reflect a collection of zones, especially when we will be transitioning between SSR Views AKA "Page Views" AKA "SPA Hard" navigations and client-side Views AKA "Interactive Views" AKA "SPA Soft" navigations.

It's probably better to keep transition start point (e.g. navigationStart & start mark) on View rather than the zone and create metrics by defining measures between transition start and last event (UserTiming mark) in the zone.

File: js/src/ExpectedMark.js (hunk 10 above)
Comment (sergeychernyshev) [reply js/src/ExpectedMark.js]: Yes, this is what I was leaning towards that, envisioned some code like this to create a zone object:
```javascript
new Zone(View)(marks)
```

File: js/src/Zone.js (hunk 7 above)
Comment (mmcgahan) [reply js/src/Zone.js:9]: ah, nice. So essentially views are comprise of zones, each zone has a set of measures it's interested in, and each measure is defined by a set of (possibly overlapping) marks.

View : Zone - one to many
Zone : Measure - one to many
Measure : Mark - many to many

I don't want to create too many pieces that we need to track, but I think having those 4 would be valuable, and we can diagram how they interact relatively easily

File: js/src/ExpectedMark.js (hunk 10 above)
Comment (mmcgahan) [reply js/src/ExpectedMark.js]: gotcha - unless a Zone needs to know its parent view, you could probably get away with something like

```js
class View {
  constructor(config) {
    const { zoneNames } = config;
    this.zones = zones.map(({ markNames }) => new Zone({ markNames }));
  }
}
```

where the parent view keeps track of Zones, but the Zone itself doesn't get any info about the view that owns it.

File: js/src/Zone.js (hunk 5 above)
Comment (sergeychernyshev) [reply js/src/Zone.js:5]: As we discussed, it is a collection of HTML elements, to some degree.

We are not doing much with it at the moment, but will use CSS selectors to find these elements.

File: js/index.html [hunk 11]
```diff
@@ -4,7 +4,7 @@
     <title>Sample Page | UX capture</title>
 
     <!-- don't do this in production, but inline ux-capture.min.js instead -->
-    <script src="ux-capture.js"></script>
+    <script src="ux-capture.min.js"></script>
```
Comment (sadafie) [inline js/index.html:7]: seems like you're using min, so remove the comment above?

File: package.json [hunk 12]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ux-capture",
-  "version": "1.0.8",
+  "version": "1.1.0",
```
Comment (sadafie) [inline package.json]: This looks like it might be a breaking change - i.e., clients are going to have to update their library usage. I'd make this a major version bumpg - `2.0.0`

File: webpack.config.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
+const webpack = require("webpack"); //to access built-in plugins
```
Comment (sadafie) [inline webpack.config.js:1]: This is a nitpick, but for future cases, it would have been nice to split the webpack changes into it's own PR/ticket and break this large PR into smaller ones. :)

File: package.json (hunk 12 above)
Comment (sergeychernyshev) [reply package.json]: I wonder if it's even worth making this change now then, because the only breaking part is renaming `label` to `name` in input structure.

File: js/index.html (hunk 11 above)
Comment (sergeychernyshev) [reply js/index.html:7]: Not exactly, inlining is still needed, just no need to point out the file name.

File: webpack.config.js (hunk 13 above)
Comment (sergeychernyshev) [reply webpack.config.js:1]: We had to rely on the bundle in this rewrite as it was impossible to run as is anymore.

ESLint could've been a separate thing though, I agree.

**Comment by sergeychernyshev** [comment]
@mmcgahan I feel I made all the changes we discussed, let me know if you if it's all set and I can merge it.

File: js/src/ExpectedMark.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
+import UXCapture from "./UXCapture";
+
+// all marks expected so far
+const expectedMarks = [];
```
Comment (mmcgahan) [inline js/src/ExpectedMark.js]: nitpick, by convention a 'private' var like this should probably get a leading underscore `const _expectedMarks`, although ESLint might complain, in which case don't worry about it.

File: js/src/ExpectedMark.js [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Class describes expected marks
+ * These marks that have to be recorded before zone is considered complete
+ */
+export default class ExpectedMark {
+  static get(name) {
+    const mark = expectedMarks.find(mark => mark.name === name);
+
+    return mark;
```
Comment (mmcgahan) [inline js/src/ExpectedMark.js:29]: I don't know if you feel like we should write this as lean as possible, but you can return directly here

```js
return expectedMarks.find(mark => mark.name === name);
```

File: js/src/ExpectedMark.js [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  constructor(name, callback) {
```
Comment (mmcgahan) [inline js/src/ExpectedMark.js]: I recommend using a single object argument `{ name, onMark }` rather than 2 separate args - easier to maintain over time and it gives the caller explicit 'named args'.

Minor - can we rename `callback` to `onMark` to make it more clear when the callback will be called?

File: js/src/ExpectedMark.js [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    this.name = name;
+
+    // list of zone callbacks to call on completion
+    this.onCompleteListeners = [];
+
+    if (!ExpectedMark.get(this.name)) {
+      expectedMarks.push(this);
+    }
```
Comment (mmcgahan) [inline js/src/ExpectedMark.js]: I think there are some subtle bugs here because the constructor will create an object that isn't actually in `expectedMarks`, and so will not be managed in the same way as the Mark with the same name that _is_ in `expectedMarks`.

I'm thinking it would be useful to define a `static create(name)` (or `static set(name)`) method rather than expecting callers to use the constructor directly. I think that would make it a little easier for the class to manage its set of `expectedMarks`. The `create` method would be able to check `expectedMarks` and conditionally instantiate a new one - that would avoid the awkward case of calling the constructor and getting back an object that isn't actually part of `expectedMarks` and is therefore kind of 'useless'.

File: js/src/UXCapture.js [hunk 18]
```diff
@@ -0,0 +1,78 @@
+import ExpectedMark from "./ExpectedMark";
+import View from "./View";
+
+export default class UXCapture {
+  static attachTo(window) {
+    // allow running in Node.js environment
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: i think with the error, we're _not_ allowing this to run in Node

File: js/src/UXCapture.js [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+
+    // assign singleton instance to window object
+    window.UX = new UXCapture(window);
+
+    return window.UX;
+  }
+
+  constructor() {
+    this.onMark = null;
+    this.onMeasure = null;
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: would we ever want these to be passed into constructor to be defined immediately?

File: js/src/UXCapture.js [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  /**
+   * Sets expected marks and corresponding zones for current view
+   *
+   * @TODO re-evaluate if we should allow multiple executions of this method
+   */
+  expect(zoneConfigs) {
+    if (typeof zoneConfigs === "undefined" || !Array.isArray(zoneConfigs)) {
+      return false;
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: sort of trivial, but the `false` return value here is a little unexpected - `expect()` will return `false` or `undefined`?  I think I would recommend printing something with `console.error()` and then `return;` with no value just to keep the return type consistent.

File: js/src/UXCapture.js [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    this.onMark = null;
+    this.onMeasure = null;
+  }
+
+  /**
+   * Sets expected marks and corresponding zones for current view
+   *
+   * @TODO re-evaluate if we should allow multiple executions of this method
+   */
+  expect(zoneConfigs) {
```
Comment (mmcgahan) [inline js/src/UXCapture.js:19]: Since this PR is a breaking change (`label`->`name`), I'll go ahead and recommend a name change here as well - instead of `expect`, which is a little vague, I wonder if this method would be easier to understand as `createView` or `initView`, which makes it more clear _what_ is being done and _when_ it should be called (i.e. as soon as a view is 'started')

File: js/src/UXCapture.js [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    // assign singleton instance to window object
+    window.UX = new UXCapture(window);
+
+    return window.UX;
+  }
+
+  constructor() {
+    this.onMark = null;
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: if these were initiallized to no-op `() => {}` functions, then you wouldn't ever have check `if (this.onMark)`

File: js/src/UXCapture.js [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+   */
+  expect(zoneConfigs) {
+    if (typeof zoneConfigs === "undefined" || !Array.isArray(zoneConfigs)) {
+      return false;
+    }
+
+    // create a view object for initial, server-side rendered page view
+    const pageView = new View({
+      onMark: mark => {
+        if (this.onMark) {
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: if `this.onMark` and `this.onMeasure` are initialized to no-ops, you can write

```js
const pageView = new View({ onMark: this.onMark, onMeasure: this.onMeasure, zoneConfigs });
```

super concise

File: js/src/UXCapture.js [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+      this.onMark = configuration.onMark;
+    } else {
+      this.onMark = null;
+    }
+
+    if (typeof configuration.onMeasure === "function") {
+      this.onMeasure = configuration.onMeasure;
+    } else {
+      this.onMeasure = null;
+    }
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: the behavior would be slightly different, but a more concise implementation would be

```js
config(configuration) {
  // destructure callback keys with default to no-op;
  const { onMark = () => {}, onMeasure = () => {} } = configuration;
  Object.assign(this, configuration);
}
```

the downside is that there is no type checking. However, I would argue that calling `config` with no arg or an arg with the wrong type is a programmer error that should `throw` - silent failures to apply a configuration will be hard to recognize and debug

File: js/src/View.js [hunk 25]
```diff
@@ -0,0 +1,18 @@
+import Zone from "./Zone";
+
+export default class View {
```
Comment (mmcgahan) [inline js/src/View.js:6]: code comment to just summarize that a `View is a collection of Zones`?

File: js/src/View.js [hunk 26]
```diff
@@ -0,0 +1,18 @@
...
+  constructor({ onMark, onMeasure, zoneConfigs }) {
+    this.expectedZones = zoneConfigs.map(zoneConfig => {
+      // only create promises if zone contains any marks, otherwise just ignore it
```
Comment (mmcgahan) [inline js/src/View.js]: no Promises

File: js/src/View.js [hunk 27]
```diff
@@ -0,0 +1,18 @@
...
+      if (zoneConfig.marks && zoneConfig.marks.length > 0) {
+        const expectedZone = new Zone({
+          onMark,
+          onMeasure,
+          ...zoneConfig
+        });
+
+        return expectedZone;
```
Comment (mmcgahan) [inline js/src/View.js]: nitpick, it's simpler to avoid assigning `expectedZone` here

```js
return new Zone({ ... });
```

File: js/src/Zone.js [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+    this.recordedMarkNames = [];
+
+    this.startMarkName = config.startMarkName || "navigationStart";
+
+    // look up existing mark object or create a new one
+
+    // now just string names for UserTiming marks
+    // in the future, different types of events, e.g. ImageElement("logo"), PaintTimer("first-paint")
+    this.marks = config.marks.map(markName => {
+      const mark = ExpectedMark.get(markName) || new ExpectedMark(markName);
```
Comment (mmcgahan) [inline js/src/Zone.js]: if we had a `ExpectedMark.create` static method, that could handle this `||` logic internally

File: js/src/Zone.js [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+
+      mark.onComplete(mark => this.checkCompletion(mark));
```
Comment (mmcgahan) [inline js/src/Zone.js]: nitpick, this can be

```js
mark.onComplete(this.checkCompletion);
```

File: js/src/Zone.js [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+   * Callback to be called by each mark in the zone upon recording
+   *
+   * @param {Mark} recordedMark latest mark recorded
+   */
+  checkCompletion(recordedMark) {
+    this.recordedMarkNames[recordedMark.name] = true;
+
+    // if callback is specified, call it with mark name
+    if (this.onMark) {
+      this.onMark(recordedMark.name);
```
Comment (mmcgahan) [inline js/src/Zone.js]: hm, this call seems like it should be made by the `ExpectedMark` in its `record` method - it would be better if `checkCompletion` was a pure function rather than triggering side effects. If `this.onMark` was passed down into the individual marks, the lifecycle responsibilities would be more clear.

File: js/src/Zone.js [hunk 31]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    // check if all marks for the zone were completed
+    const isComplete = this.marks.every(
+      mark => this.recordedMarkNames[mark.name]
+    );
+
+    if (isComplete) {
```
Comment (mmcgahan) [inline js/src/Zone.js]: definitely nitpick, but I think it would be a simple update:

Let's separate `checkCompletion` from _handling_ completion, e.g. define an `onComplete` method for `Zone` that would encapsulate the 'complete' behavior independent of checking whether the zone is complete

The `mark.onComplete` call would then look like

```js
mark.onComplete(mark => {
  if (this.checkCompletion(mark)) { this.onComplete(); }
});
```

which is a little more verbose, but keeps the `onComplete` lifecycle similar between Mark and Zone

File: js/src/Zone.js (hunk 29 above)
Comment (sergeychernyshev) [reply js/src/Zone.js]: Not really as `checkCompletion` is method of `Zone`, but `mark.record()` method would be calling it in it's own context where `this` inside that function will be referring to mark object instead.

Comment (sergeychernyshev) [reply js/src/Zone.js]: In fact, now it handles checking and calls Zone's onComplete too so no shortcutting ;)

File: js/src/ExpectedMark.js (hunk 16 above)
Comment (sergeychernyshev) [reply js/src/ExpectedMark.js]: Callbacks are actually registered using onComplete() now so callback param is not necessary

File: js/src/UXCapture.js (hunk 20 above)
Comment (sergeychernyshev) [reply js/src/UXCapture.js]: Will be throwing an error here, similar to how we do that for a window - there is no point to silently accept `UX.expect()` call without an array passed in.

It's OK to pass an empty array though.

File: js/src/UXCapture.js (hunk 19 above)
Comment (sergeychernyshev) [reply js/src/UXCapture.js]: Not really, because it gets constructed automatically by the library when it attaches it to the window and then user of the library configures it later.

File: js/src/UXCapture.js (hunk 21 above)
Comment (sergeychernyshev) [reply js/src/UXCapture.js:19]: It makes sense, but I would delay this change to next major version when we have SPA support when we'll really need multiple views - in this change View is just simplification of the library logic rather than really functional element.

At that point we will need need new API for supplying expected marks for multiple views (see https://meetup.atlassian.net/browse/WP-829)

The question comes though if we should really do `label` -> `name` change now and increment major to 2.x.x at all then?

File: js/src/UXCapture.js (hunk 23 above)
Comment (sergeychernyshev) [reply js/src/UXCapture.js]: I kept the closures to make sure we can call `UX.config()` after `UX.expect()` was called.

Not a big deal, but I felt that it is a useful simplification that removes dependency on sequence of execution.

File: js/src/UXCapture.js [hunk 32]
```diff
@@ -0,0 +1,74 @@
+import ExpectedMark from "./ExpectedMark";
+import View from "./View";
+
+const NOOP = () => {};
+
+export default class UXCapture {
+  static attachTo(window) {
+    // require a valid window object to be passed in
+    if (typeof window === "undefined") {
+      throw new Error("Must provide a valid window object");
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: if `window` is undefined, the check for `window.UX` on line 14 will throw `window is not defined`, so I think we should avoid the separate manual check to keep the code lean.

_Not a blocker, but would be good for followup_

File: js/src/UXCapture.js [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    // already have UX Capture object defined, reuse it
+    if (typeof window.UX !== "undefined") {
+      return window.UX;
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: a relatively common convention for this kind of default assignment is

```js
window.UX = window.UX || new UXCapture();
```

_Not a blocker, but would be good for followup_

File: js/src/UXCapture.js [hunk 34]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    // assign singleton instance to window object
+    window.UX = new UXCapture(window);
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: constructor no longer takes `window` arg

File: js/src/UXCapture.js (hunk 21 above)
Comment (mmcgahan) [reply js/src/UXCapture.js:19]: up to you - I feel like this is a big enough rewrite to warrant its own major version, and I personally don't mind packages having very high version numbers, so I don't mind if breaking changes are spaced out.

File: js/src/UXCapture.js [hunk 35]
```diff
@@ -0,0 +1,74 @@
...
+
+  /**
+   * Sets expected marks and corresponding zones for current view
+   *
+   * @TODO re-evaluate if we should allow multiple executions of this method
+   */
+  expect(zoneConfigs) {
+    if (typeof zoneConfigs === "undefined" || !Array.isArray(zoneConfigs)) {
+      throw new Error("Must provide an array of zone configuration objects");
+    }
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: Do we need this type checking? the `View` constructor will throw an error if `zoneConfigs` is not an array, so I think we can avoid this incidental complexity in `UXCapture`

File: js/src/UXCapture.js [hunk 36]
```diff
@@ -0,0 +1,74 @@
+import ExpectedMark from "./ExpectedMark";
+import View from "./View";
+
+const NOOP = () => {};
+
+export default class UXCapture {
+  static attachTo(window) {
```
Comment (mmcgahan) [inline js/src/UXCapture.js]: can we call this method `init` or `create`? Backbone used `init` for initializing applications, and since we don't actually need to pass in `window`, probably better to use slightly more common terminology

_Not a blocker, but would be good for followup_

File: js/src/Zone.js [hunk 37]
```diff
@@ -0,0 +1,91 @@
...
+    this.startMarkName = config.startMarkName || "navigationStart";
+
+    // look up existing mark object or create a new one
+
+    // now just string names for UserTiming marks
+    // in the future, different types of events, e.g. ImageElement("logo"), PaintTimer("first-paint")
+    this.marks = config.marks.map(markName => {
+      const mark = ExpectedMark.create(markName);
+
+      mark.onComplete(mark => {
```
Comment (mmcgahan) [inline js/src/Zone.js:39]: React encourages a pattern that I think would be helpful here - instead of imperatively adding listeners to the `mark` with an instance method after it's been instantiated, the listener can instead be passed in declaratively as an `onMark` 'prop' that gets assigned/merged into the mark's `onMarkListeners`. It wouldn't affect the top level API, but it would simplify the `ExpectedMark` interface, and would parallel the `onMark`/`onMeasure` callback passing interface that `Zone` has, which is good for consistency/understandability, particularly for React developers.

The main interface change is that `name` and `onMark` would both have to be passed into the `ExpectedMark` `create` method and constructor

```js
// Zone constructor
// compose `this.onMark` from `config.onMark` _plus additional 'complete' logic_
this.onMark = (markName => {
  config.onMark(markName);
  this.checkCompletion(markName) && this.complete(markName);
}).bind(this);

// `this.marks` assignment becomes a simpler mapping, with each `ExpectedMark` getting a `this.onMark` callback
this.marks = config.marks.map(name => ExpectedMark.create({ name, onMark: this.onMark }));
```

`ExpectedMark` would update to manage `this.onMarkListeners` directly in the constructor and `create` static methods

```js
static create(config) {
  let mark = ExpectedMark.get(config.name);
  if (mark) {
    mark.onMarkListeners.push(config.onMark || NOOP);
  } else {
    mark = new ExpectedMark(config);
    _expectedMarks.push(mark);
  }
   return mark;
}

constructor(config) {
  this.name = config.name;
  this.onMarkListeners = [config.onMark || NOOP];
}
```

_Not a blocker, but would be good for followup_

File: js/src/Zone.js [hunk 38]
```diff
@@ -0,0 +1,91 @@
...
+        // if Zone's onMark callback is specified, call it with mark name
+        if (this.onMark) {
+          this.onMark(mark.name);
+        }
+
+        if (this.checkCompletion(mark)) {
```
Comment (mmcgahan) [inline js/src/Zone.js]: not sure if it's clear in my suggestion above, but you can merge this behavior into the `this.onMark` assignment on line 28 to make the code simpler

_Not a blocker, but would be good for followup_

**Comment by mmcgahan** [comment]
I've edited my comments to clarify the ones that are not blockers - I know you're eager to get this update published as long as we don't lose some optimization TODOs

File: js/src/UXCapture.js (hunk 21 above)
Comment (sergeychernyshev) [reply js/src/UXCapture.js:19]: I'll leave createView to the moment when we have SPA implementation to make sure we have a good API there rather than rushing it in now, but will keep 2.x increment with `label`->`name` renaming in place.

File: js/src/UXCapture.js (hunk 24 above)
Comment (sergeychernyshev) [reply js/src/UXCapture.js]: Well, assign shouldn't use configuration object itself, but rather the onMark and onMeasure with defaults, but I got the point ;)

File: js/src/UXCapture.js (hunk 34 above)
Comment (sergeychernyshev) [reply js/src/UXCapture.js]: Removed attachTo() altogether as we now use global window object and it doesn't make sense to wrap it.

File: js/src/UXCapture.js (hunk 36 above)
Comment (sergeychernyshev) [reply js/src/UXCapture.js]: Nuked it altogether, window.UX is attached right away in the browser and in Node you'll have to run new on it as with anything else.

File: js/src/Zone.js (hunk 37 above)
Comment (sergeychernyshev) [reply js/src/Zone.js:39]: Let's do this in a followup. I am worried also that we have external API with in `UX.config({onMark, onMeasure})` that is different from internal `onMark` callback on `ExpectedMark` object and it might be confusing the use-case here.

File: js/src/Zone.js (hunk 38 above)
Comment (sergeychernyshev) [reply js/src/Zone.js]: See my comment above about potentially confusing API.

File: js/src/Zone.js (hunk 37 above)
Comment (sergeychernyshev) [reply js/src/Zone.js:39]: I'll create a ticket for this.

File: js/src/Zone.js (hunk 5 above)
Comment (jaclynj) [reply js/src/Zone.js:5]: maybe explicitly say 'HTML element' then?

File: js/src/Zone.js [hunk 39]
```diff
@@ -0,0 +1,91 @@
...
+      return mark;
+    });
+  }
+
+  /**
+   * Callback to be called by each mark in the zone upon recording
+   *
+   * @param {Mark} recordedMark latest mark recorded
+   */
+  checkCompletion(recordedMark) {
```
Comment (jaclynj) [inline js/src/Zone.js]: This is a little confusing to me - 
This method does two things, 1. it marks the latest `recordedMark` as 'complete' and 2. it checks the completion of the other `recordedMarkNames` right?

Nitpick - I would expect `checkCompletion` to only do the second thing.
Question - Why is `recordedMarkNames` an array?

**Review by jaclynj — COMMENTED** [review]
Looks great! Code is clear and easy to understand. Added 2 small non-blocking comments.

File: js/src/Zone.js (hunk 39 above)
Comment (sergeychernyshev) [reply js/src/Zone.js]: Good catch, it's supposed to be an object if we want to use key uniqueness here, you know what, I should probably rewrite this to check UserTiming API instead of internal function.

Comment (sergeychernyshev) [reply js/src/Zone.js]: Rewritten.

Comment (jaclynj) [reply js/src/Zone.js]: 👍
